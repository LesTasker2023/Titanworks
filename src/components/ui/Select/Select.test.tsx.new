import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from './index';

describe('Select', () => {
  const renderBasicSelect = (selectProps = {}, triggerProps = {}) => {
    return render(
      <Select data-testid="select" {...selectProps}>
        <SelectTrigger data-testid="select-trigger" {...triggerProps}>
          <SelectValue placeholder="Select an option" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="option1">Option 1</SelectItem>
          <SelectItem value="option2">Option 2</SelectItem>
        </SelectContent>
      </Select>
    );
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(
        <Select>
          <SelectTrigger>
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="option1">Option 1</SelectItem>
          </SelectContent>
        </Select>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches loading state snapshot', () => {
      const { container } = render(
        <Select>
          <SelectTrigger loading>
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="option1">Option 1</SelectItem>
          </SelectContent>
        </Select>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches disabled state snapshot', () => {
      const { container } = render(
        <Select disabled>
          <SelectTrigger>
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="option1">Option 1</SelectItem>
          </SelectContent>
        </Select>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches error state snapshot', () => {
      const { container } = render(
        <Select>
          <SelectTrigger error>
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="option1">Option 1</SelectItem>
          </SelectContent>
        </Select>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', () => {
      renderBasicSelect();
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('shows placeholder text', () => {
      renderBasicSelect();
      expect(screen.getByText('Select an option')).toBeInTheDocument();
    });
  });

  describe('States', () => {
    it('handles loading state correctly', () => {
      renderBasicSelect({}, { loading: true });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles disabled state correctly', () => {
      renderBasicSelect({ disabled: true });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles error state correctly', () => {
      renderBasicSelect({}, { error: true });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      renderBasicSelect();
      const trigger = screen.getByTestId('select-trigger');
      expect(trigger).toHaveAttribute('role', 'combobox');
    });

    it('announces changes to screen readers', () => {
      renderBasicSelect();
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicSelect();
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className on trigger', () => {
      renderBasicSelect({}, { className: 'custom-class' });
      const trigger = screen.getByTestId('select-trigger');
      expect(trigger).toHaveClass('custom-class');
    });

    it('forwards refs correctly', () => {
      const ref = vi.fn();
      renderBasicSelect({ ref });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('spreads additional props on trigger', () => {
      renderBasicSelect({}, { 'data-custom': 'test-value' });
      const trigger = screen.getByTestId('select-trigger');
      expect(trigger).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicSelect({ children: undefined });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicSelect({ children: null });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicSelect({}, { className: '' });
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles rapid prop changes', () => {
      const { rerender } = renderBasicSelect({}, { className: 'class1' });
      rerender(
        <Select data-testid="select">
          <SelectTrigger data-testid="select-trigger" className="class2">
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="option1">Option 1</SelectItem>
          </SelectContent>
        </Select>
      );
      const trigger = screen.getByTestId('select-trigger');
      expect(trigger).toHaveClass('class2');
    });

    it('handles complex nested content', () => {
      render(
        <Select data-testid="select">
          <SelectTrigger data-testid="select-trigger">
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="complex">
              <div>
                <span>Complex</span>
                <p>Nested content</p>
              </div>
            </SelectItem>
          </SelectContent>
        </Select>
      );
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <Select data-testid="select">
          <SelectTrigger data-testid="select-trigger">
            <SelectValue placeholder="Select an option" />
          </SelectTrigger>
          <SelectContent>
            {Array.from({ length: 100 }, (_, i) => (
              <SelectItem key={i} value={`option${i}`}>Option {i}</SelectItem>
            ))}
          </SelectContent>
        </Select>
      );
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicSelect();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicSelect();
      unmount();
      renderBasicSelect();
      expect(screen.getByTestId('select-trigger')).toBeInTheDocument();
    });
  });
});
