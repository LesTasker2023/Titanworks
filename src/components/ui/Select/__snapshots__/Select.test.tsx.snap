// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > Snapshots > matches default snapshot 1`] = `
<button
  aria-autocomplete="none"
  aria-controls="radix-«r3»"
  aria-expanded="false"
  class="select__trigger"
  data-placeholder=""
  data-state="closed"
  dir="ltr"
  role="combobox"
  type="button"
>
  <span
    style="pointer-events: none;"
  >
    Select an option
  </span>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</button>
`;

exports[`Select > Snapshots > matches disabled state snapshot 1`] = `
<button
  aria-autocomplete="none"
  aria-controls="radix-«rd»"
  aria-expanded="false"
  class="select__trigger"
  data-disabled=""
  data-placeholder=""
  data-state="closed"
  dir="ltr"
  disabled=""
  role="combobox"
  type="button"
>
  <span
    style="pointer-events: none;"
  >
    Select an option
  </span>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</button>
`;

exports[`Select > Snapshots > matches error state snapshot 1`] = `
<button
  aria-autocomplete="none"
  aria-controls="radix-«ri»"
  aria-expanded="false"
  class="select__trigger select__trigger--error"
  data-placeholder=""
  data-state="closed"
  dir="ltr"
  role="combobox"
  type="button"
>
  <span
    style="pointer-events: none;"
  >
    Select an option
  </span>
  <svg
    aria-hidden="true"
    class="lucide lucide-chevron-down select__icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m6 9 6 6 6-6"
    />
  </svg>
</button>
`;

exports[`Select > Snapshots > matches loading state snapshot 1`] = `
<button
  aria-autocomplete="none"
  aria-controls="radix-«r8»"
  aria-expanded="false"
  class="select__trigger select__trigger--loading"
  data-disabled=""
  data-placeholder=""
  data-state="closed"
  dir="ltr"
  disabled=""
  role="combobox"
  type="button"
>
  <span
    style="pointer-events: none;"
  >
    Select an option
  </span>
  <div
    aria-hidden="true"
    class="select__spinner"
  />
</button>
`;
